p2p/src/supervisor.rs

```diff
@@ -0,0 +1,279 @@
+}
+
+enum Output<Conn> {
+    Event(Event),
+    Internal(Internal<Conn>),
+}
+
+impl<Conn> From<Event> for Output<Conn> {
+    fn from(event: Event) -> Self {
+        Output::Event(event)
+    }
+}
+
+impl<Conn> From<Internal<Conn>> for Output<Conn> {
+    fn from(internal: Internal<Conn>) -> Self {
+        Output::Internal(internal)
+    }
+}
+
+pub struct Supervisor {
+    command: Sender<Command>,
+    events: Receiver<Event>,
+}
+
+impl Supervisor {
+    pub fn run<T>(transport: T) -> Result<Self>
+    where
+        T: transport::Transport + Send + 'static,
+    {
+        let (command, commands) = unbounded();
+        let (event_tx, events) = unbounded();
+        let supervisor = Self { command, events };
+
+        let (endpoint, mut incoming) = transport.bind(transport::BindInfo {
+            addr: SocketAddr::new(IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)), 12345),
+            advertise_addrs: vec![SocketAddr::new(
+                IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)),
+                12345,
+            )],
+            public_key: PublicKey::from_raw_ed25519(&[
+                215, 90, 152, 1, 130, 177, 10, 183, 213, 75, 254, 211, 201, 100, 7, 58, 14, 225,
+                114, 243, 218, 166, 35, 37, 175, 2, 26, 104, 247, 7, 81, 26,
+            ])
+            .unwrap(),
+        })?;
+
+        // ACCEPT
+        let (accept_tx, accept_rx) = unbounded::<()>();
+        let (accepted_tx, accepted_rx) = unbounded();
+        thread::spawn(move || loop {
+            accept_rx.recv().unwrap();
+
+            let conn = incoming.next().unwrap().unwrap();
+            let id = match conn.public_key() {
+                PublicKey::Ed25519(ed25519) => node::Id::from(ed25519),
+                _ => panic!(),
+            };
+
+            accepted_tx.send(Input::Accepted(id, conn)).unwrap();
+        });
+
+        // CONNECT
+        let (_connect_tx, connect_rx) = unbounded::<transport::ConnectInfo>();
+        let (connected_tx, connected_rx) = unbounded();
+        thread::spawn(move || loop {
+            let info = connect_rx.recv().unwrap();
+            let conn = endpoint.connect(info).unwrap();
+            let id = match conn.public_key() {
+                PublicKey::Ed25519(ed25519) => node::Id::from(ed25519),
+                _ => panic!(),
+            };
+
+            connected_tx.send(Input::Connected(id, conn)).unwrap();
+        });
+
+        // UPGRADE
+        let (upgrade_tx, upgrade_rx) =
+            unbounded::<transport::Direction<<T as transport::Transport>::Connection>>();
+        let (upgraded_tx, upgraded_rx) = unbounded();
+        thread::spawn(move || loop {
+            let conn = upgrade_rx.recv().unwrap();
+            let peer = peer::Peer::from(conn);
+
+            upgraded_tx
+                .send(Input::Upgraded(peer.id, peer.run(vec![])))
+                .unwrap();
+        });
+
+        // MAIN
+        thread::spawn(move || {
```

Comment:
I think I understand what you're achieving here with spawning the main supervisor loop as a separate thread and then returning the supervisor instance (i.e. so you can send it commands from the initiating thread), but how does the owner know if the supervisor's main thread has panicked and is no longer available? (e.g. there are a bunch of `unwrap`s in the loop)

Reply:
Perhaps we could refactor this so that the `run` method returns a `Result` and leave it to the caller to spawn a thread for the supervisor itself, much like we do [in the light client](https://github.com/informalsystems/tendermint-rs/blob/3c7754b98e09d965973bf6580d48dcce9216e022/light-client/src/supervisor.rs#L330).

This would require splitting the `Supervisor` struct into an actual `Supervisor` with a `run` method, and a clonable `SupervisorHandle` which can then be use to send commands to the supervisor.

```rust
impl Supervisor {
  // option 1
  pub fn new() -> Self;
  pub fn handle(&self) -> SupervisorHandle;
  // option 2
  pub fn new() -> (Self, SupervisorHandle);

  // common
  pub fn run(self) -> Result<(), SupervisorError>;
}

#[derive(Clone)]
struct SupervisorHandle {
    command: Sender<Command>,
    events: Receiver<Event>,
}
```

Reply:
The extra indirection of the handle does not add much value here as the returned `Supervisor` acts as such. The `uwrap`s will be refactored out and the public interface will indicate if the internal machinery dissolved/imploded.

Beyond that there doesn't seem to be some gain in having extra lifecycle control on the caller side, a transport is already passed in with the intent to be bound and managed. It is questionable what could happen between the construction and explicit call of `run` that wold be meaningful. In turn it could lead to unexpected behaviour, when the run is never spawned and the transport is bound and its resources are locked up.